### DISCUSSION TITLE
What happens to the FP32 biases in HardWare ?

### DISCUSSION THREAD

**Discussion by 0BAB1** [Xilinx/finn#1181] [Q&A] [answered]
In [this notebook](https://github.com/Xilinx/finn/blob/main/notebooks/advanced/4_advanced_builder_settings.ipynb), you specify that FINN only deals with integers :

![image](https://github.com/user-attachments/assets/c3639f25-4c1a-4059-a9c8-e89a506a1394)

The thing is, when working with brevitas,a likely scenario (the one I use for example) is that we do not quantize the Biases, making them floating point numbers.

I some scenario we even only quantize the weights (Event though I did not try that) : https://xilinx.github.io/brevitas/getting_started.html

I think I may be confused with the training part that introduces Quant error for the model to adapt ? but when i observe the biases on a quantized model, turns out they are FP32...

So what happens to the biases ? Is FP32 used between the layer and then quantized through thresholding ? 

I would love to understand this point to deepen my understanding of the tool, thanks in advance for all the insights.

**Comment by auphelia** [comment, marked as answer]
Hi @0BAB1,
In the streamlining step in FINN, we move floating point operations in the graph to reveal the underlying integer compute. This process is based on this paper: https://arxiv.org/abs/1709.04060 and preserves the functional behavior.
You can find some more information in the notebooks: [tfc_end2end_example.ipynb](https://github.com/Xilinx/finn/blob/dev/notebooks/end2end_example/bnn-pynq/tfc_end2end_example.ipynb) and [cnv_end2end_example.ipynb](https://github.com/Xilinx/finn/blob/dev/notebooks/end2end_example/bnn-pynq/cnv_end2end_example.ipynb). 
Additionally, you can have a look at the transformations yourself: https://github.com/Xilinx/finn/tree/dev/src/finn/transformation/streamline

**Reply by 0BAB1** [reply]
Thanks for all the valuable ressources !

The notebooks have a very high level approach, thus the question.
